src/cls/async-hooks.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as asyncHooksModule from 'async_hooks';
+import {EventEmitter} from 'events';
+import * as shimmer from 'shimmer';
+
+import {CLS, Func} from './base';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> =
+    ['addListener', 'on', 'once', 'prependListener', 'prependOnceListener'];
+const WRAPPED = Symbol('context_wrapped');
+
+/**
+ * An implementation of continuation-local storage on top of the async_hooks
+ * module.
+ */
+export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
+  private current: {value: Context};
```

Comment:
nit: would `currentContext` be more descriptive?

Reply:
Done